cmd/juicify/app/juicify.go

```diff
@@ -5,66 +5,151 @@ package app
+}
+
+var (
+	controller  = flag.String("controller", "", "The IP address or hostname and port of the controller")
+	accessToken = flag.String("access-token", "", "The access token to use when connecting to the controller")
+
+	test           = flag.Bool("test", false, "Deprecated: Use --test-connection instead")
+	testConnection = flag.Bool("test-connection", false, "Tests the reachability of the controller or server(s)")
+
+	queueTimeout      = flag.Uint("queue-timeout", 0, "Maximum number of seconds to wait for a GPU")
+	onQueueTimeout    = flag.String("on-queue-timeout", "fail", "When a queue timeout happens, [fail, continue]")
+	onConnectionError = flag.String("on-connection-error", "fail", "When a connection error happens, [fail, continue]")
+
+	juicePath = flag.String("juice-path", "", "Path to the juice executables if different than current executable path")
+
+	errCancelled            = errors.New("cancelled")
+	errQueueTimeout         = errors.New("queued GPU request timed out")
+	errInvalidConfiguration = errors.New("invalid configuration")
+)
+
+func validateConfiguration(config *Configuration) error {
+	if config.Id != "" {
+		return errInvalidConfiguration.Wrap(errors.New("id must not be specified"))
+	} else if len(config.Servers) != 0 {
+		return errInvalidConfiguration.Wrap(errors.New("servers must not be specified"))
+	} else if config.Controller == "" {
+		return errInvalidConfiguration.Wrap(errors.New("controller must be specified"))
+	} else if len(config.Requirements.Gpus) == 0 {
+		config.Requirements.Gpus = append(config.Requirements.Gpus, restapi.GpuRequirements{})
+	}
+
+	return nil
+}
+
+func requestSession(group task.Group, client *http.Client, config *Configuration) error {
+	api := restapi.Client{
+		Client:      client,
+		Address:     config.Controller,
+		AccessToken: config.AccessToken,
+	}
+
+	id, err := api.RequestSessionWithContext(group.Ctx(), config.Requirements)
+	if err != nil {
+		return err
+	}
+
+	session, err := api.GetSessionWithContext(group.Ctx(), id)
+	if err != nil {
+		return err
+	}
 
-	LogGroup string `json:"logGroup,omitempty"`
+	if session.State == restapi.SessionQueued {
+		logger.Info("Session queued")
 
-	LogFile string `json:"logFile,omitempty"`
+		timeoutChannel := make(chan struct{})
+		defer close(timeoutChannel)
 
-	ForceSoftwareDecode bool `json:"forceSoftwareDecode,omitempty"`
+		if *queueTimeout > 0 {
+			group.GoFn("Queue Timeout", func(g task.Group) error {
+				timeout := time.NewTicker(time.Duration(*queueTimeout) * time.Second)
+				defer timeout.Stop()
 
-	DisableCache bool `json:"disableCache,omitempty"`
+				select {
+				case <-group.Ctx().Done():
+					return nil
 
-	DisableCompression bool `json:"disableCompression,omitempty"`
+				case <-timeout.C:
+					timeoutChannel <- struct{}{}
+					return nil
+				}
+			})
+		}
 
-	Headless bool `json:"headless,omitempty"`
+		ticker := time.NewTicker(5 * time.Second)
```

Comment:
should we poll more frequently than this?

Reply:
All the explicit values should be configurable in some way. This was just a reasonable value.